### ISSUE TITLE
It seams that the DIRE tensor save format: jpg or png, determine the results of the resnet50 detector

### ISSUE THREAD

**Issue by JYccode** [ZhendongWang6/DIRE#30]
my computh_dir.sh is
```
## set MODEL_PATH, num_samples, has_subfolder, images_dir, recons_dir, dire_dir
export CUDA_VISIBLE_DEVICES=0
export NCCL_P2P_DISABLE=1
MODEL_PATH="../models/256x256_diffusion_uncond.pt" # "models/lsun_bedroom.pt, models/256x256_diffusion_uncond.pt"

SAMPLE_FLAGS="--batch_size 1 --num_samples 4  --timestep_respacing ddim20 --use_ddim True"
SAVE_FLAGS="--images_dir ../data/single_test --recons_dir ../recons_test/single_test --dire_dir ../dire_test/single_test"
MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond False --diffusion_steps 1000 --dropout 0.1 --image_size 256 --learn_sigma True --noise_schedule linear --num_channels 256 --num_head_channels 64 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"
mpiexec --allow-run-as-root -n 1 python compute_dire.py --model_path $MODEL_PATH $MODEL_FLAGS  $SAVE_FLAGS $SAMPLE_FLAGS --has_subfolder True
```
the diffusion model is 256x256_diffusion_uncond.pt, but i also tried other models like lsun_bedroom.pt.
then I run computh_dir.py to get the DIRE img. 
then I run the demo.py to use resnet50 cnn model which weights is lsun_adm.pt:
```
python demo.py -f /data/github_issue/DIRE/dire_test/single_test/single_test -m /data/github_issue/DIRE/models/lsun_adm.pth
```
this scripts can get Prob of being synthetic.
In the test, the Fake image is png format, the real image is jpg format. these image is download from DiffusionForensics dataasets
My question is: when I using computh_dir.py to save the DIRE tensor to "PNG" format, the Prob of being synthetic always 1.0000; In the other hand, save to "JPG" format, the Prob of being synthetic always 0.0000, no matter whether fake or real image i use.

**Comment by JYccode** [comment]
additionally, the only modify of the code is the compute_dire.py:
```
fn_save = fn_save.replace("png", "jpg")
```
add the line constrain to save the dire tensor to "png" format image file

**Comment by bGN5** [comment]
> 我的computh_dir.sh是
> 
> ```
> ## set MODEL_PATH, num_samples, has_subfolder, images_dir, recons_dir, dire_dir
> export CUDA_VISIBLE_DEVICES=0
> export NCCL_P2P_DISABLE=1
> MODEL_PATH="../models/256x256_diffusion_uncond.pt" # "models/lsun_bedroom.pt, models/256x256_diffusion_uncond.pt"
> 
> SAMPLE_FLAGS="--batch_size 1 --num_samples 4  --timestep_respacing ddim20 --use_ddim True"
> SAVE_FLAGS="--images_dir ../data/single_test --recons_dir ../recons_test/single_test --dire_dir ../dire_test/single_test"
> MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond False --diffusion_steps 1000 --dropout 0.1 --image_size 256 --learn_sigma True --noise_schedule linear --num_channels 256 --num_head_channels 64 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"
> mpiexec --allow-run-as-root -n 1 python compute_dire.py --model_path $MODEL_PATH $MODEL_FLAGS  $SAVE_FLAGS $SAMPLE_FLAGS --has_subfolder True
> ```
> 
> 扩散模型是 256x256_diffusion_uncond.pt，但我也尝试了其他模型，如 lsun_bedroom.pt。然后我computh_dir.py跑去获取夜魇 img。然后我运行 demo.py 以使用 resnet50 cnn 模型，其权重为 lsun_adm.pt：
> 
> ```
> python demo.py -f /data/github_issue/DIRE/dire_test/single_test/single_test -m /data/github_issue/DIRE/models/lsun_adm.pth
> ```
> 
> 此脚本可能会获得 Prob 的合成性。在测试中，假图为 png 格式，实图为 jpg 格式。这些图像是从 DiffusionForensics 数据集下载的 我的问题是：当我使用 computh_dir.py 将 DIRE 张量保存为“PNG”格式时，合成的概率始终为 1.0000;另一方面，保存为“JPG”格式，无论我使用的是假图像还是真图像，合成的可能性始终为 0.0000。

hello,i have trouble to use one gpu to run the compute-dire.py.only chnage CUDA_VISIBLE_DEVICES to 0 ,can it run on one gpu device? if you could answer me i couldn't be more apprecuated thanks

**Comment by lcayvinliu** [comment]
> > 我的computh_dir.sh是
> > ```
> > ## set MODEL_PATH, num_samples, has_subfolder, images_dir, recons_dir, dire_dir
> > export CUDA_VISIBLE_DEVICES=0
> > export NCCL_P2P_DISABLE=1
> > MODEL_PATH="../models/256x256_diffusion_uncond.pt" # "models/lsun_bedroom.pt, models/256x256_diffusion_uncond.pt"
> > 
> > SAMPLE_FLAGS="--batch_size 1 --num_samples 4  --timestep_respacing ddim20 --use_ddim True"
> > SAVE_FLAGS="--images_dir ../data/single_test --recons_dir ../recons_test/single_test --dire_dir ../dire_test/single_test"
> > MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond False --diffusion_steps 1000 --dropout 0.1 --image_size 256 --learn_sigma True --noise_schedule linear --num_channels 256 --num_head_channels 64 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"
> > mpiexec --allow-run-as-root -n 1 python compute_dire.py --model_path $MODEL_PATH $MODEL_FLAGS  $SAVE_FLAGS $SAMPLE_FLAGS --has_subfolder True
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 扩散模型是 256x256_diffusion_uncond.pt，但我也尝试了其他模型，如 lsun_bedroom.pt。然后我computh_dir.py跑去获取夜魇 img。然后我运行 demo.py 以使用 resnet50 cnn 模型，其权重为 lsun_adm.pt：
> > ```
> > python demo.py -f /data/github_issue/DIRE/dire_test/single_test/single_test -m /data/github_issue/DIRE/models/lsun_adm.pth
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 此脚本可能会获得 Prob 的合成性。在测试中，假图为 png 格式，实图为 jpg 格式。这些图像是从 DiffusionForensics 数据集下载的 我的问题是：当我使用 computh_dir.py 将 DIRE 张量保存为“PNG”格式时，合成的概率始终为 1.0000;另一方面，保存为“JPG”格式，无论我使用的是假图像还是真图像，合成的可能性始终为 0.0000。
> 
> hello,i have trouble to use one gpu to run the compute-dire.py.only chnage CUDA_VISIBLE_DEVICES to 0 ,can it run on one gpu device? if you could answer me i couldn't be more apprecuated thanks

I can run the script on one gpu device, could you provide the runtime error info in detail?

**Comment by bGN5** [comment]
> > > 我的computh_dir.sh是
> > > ```
> > > ## set MODEL_PATH, num_samples, has_subfolder, images_dir, recons_dir, dire_dir
> > > export CUDA_VISIBLE_DEVICES=0
> > > export NCCL_P2P_DISABLE=1
> > > MODEL_PATH="../models/256x256_diffusion_uncond.pt" # "models/lsun_bedroom.pt, models/256x256_diffusion_uncond.pt"
> > > 
> > > SAMPLE_FLAGS="--batch_size 1 --num_samples 4  --timestep_respacing ddim20 --use_ddim True"
> > > SAVE_FLAGS="--images_dir ../data/single_test --recons_dir ../recons_test/single_test --dire_dir ../dire_test/single_test"
> > > MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond False --diffusion_steps 1000 --dropout 0.1 --image_size 256 --learn_sigma True --noise_schedule linear --num_channels 256 --num_head_channels 64 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"
> > > mpiexec --allow-run-as-root -n 1 python compute_dire.py --model_path $MODEL_PATH $MODEL_FLAGS  $SAVE_FLAGS $SAMPLE_FLAGS --has_subfolder True
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 扩散模型是 256x256_diffusion_uncond.pt，但我也尝试了其他模型，如 lsun_bedroom.pt。然后我computh_dir.py跑去获取夜魇 img。然后我运行 demo.py 以使用 resnet50 cnn 模型，其权重为 lsun_adm.pt：
> > > ```
> > > python demo.py -f /data/github_issue/DIRE/dire_test/single_test/single_test -m /data/github_issue/DIRE/models/lsun_adm.pth
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 此脚本可能会获得 Prob 的合成性。在测试中，假图为 png 格式，实图为 jpg 格式。这些图像是从 DiffusionForensics 数据集下载的 我的问题是：当我使用 computh_dir.py 将 DIRE 张量保存为“PNG”格式时，合成的概率始终为 1.0000;另一方面，保存为“JPG”格式，无论我使用的是假图像还是真图像，合成的可能性始终为 0.0000。
> > 
> > hello,i have trouble to use one gpu to run the compute-dire.py.only chnage CUDA_VISIBLE_DEVICES to 0 ,can it run on one gpu device? if you could answer me i couldn't be more apprecuated thanks
> 
> I can run the script on one gpu device, could you provide the runtime error info in detail?

i am in class. But i do experiment on windows ,did you do experiment on linux?

**Comment by lcayvinliu** [comment]
> > > > 我的computh_dir.sh是
> > > > ```
> > > > ## set MODEL_PATH, num_samples, has_subfolder, images_dir, recons_dir, dire_dir
> > > > export CUDA_VISIBLE_DEVICES=0
> > > > export NCCL_P2P_DISABLE=1
> > > > MODEL_PATH="../models/256x256_diffusion_uncond.pt" # "models/lsun_bedroom.pt, models/256x256_diffusion_uncond.pt"
> > > > 
> > > > SAMPLE_FLAGS="--batch_size 1 --num_samples 4  --timestep_respacing ddim20 --use_ddim True"
> > > > SAVE_FLAGS="--images_dir ../data/single_test --recons_dir ../recons_test/single_test --dire_dir ../dire_test/single_test"
> > > > MODEL_FLAGS="--attention_resolutions 32,16,8 --class_cond False --diffusion_steps 1000 --dropout 0.1 --image_size 256 --learn_sigma True --noise_schedule linear --num_channels 256 --num_head_channels 64 --num_res_blocks 2 --resblock_updown True --use_fp16 True --use_scale_shift_norm True"
> > > > mpiexec --allow-run-as-root -n 1 python compute_dire.py --model_path $MODEL_PATH $MODEL_FLAGS  $SAVE_FLAGS $SAMPLE_FLAGS --has_subfolder True
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 扩散模型是 256x256_diffusion_uncond.pt，但我也尝试了其他模型，如 lsun_bedroom.pt。然后我computh_dir.py跑去获取夜魇 img。然后我运行 demo.py 以使用 resnet50 cnn 模型，其权重为 lsun_adm.pt：
> > > > ```
> > > > python demo.py -f /data/github_issue/DIRE/dire_test/single_test/single_test -m /data/github_issue/DIRE/models/lsun_adm.pth
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 此脚本可能会获得 Prob 的合成性。在测试中，假图为 png 格式，实图为 jpg 格式。这些图像是从 DiffusionForensics 数据集下载的 我的问题是：当我使用 computh_dir.py 将 DIRE 张量保存为“PNG”格式时，合成的概率始终为 1.0000;另一方面，保存为“JPG”格式，无论我使用的是假图像还是真图像，合成的可能性始终为 0.0000。
> > > 
> > > 
> > > hello,i have trouble to use one gpu to run the compute-dire.py.only chnage CUDA_VISIBLE_DEVICES to 0 ,can it run on one gpu device? if you could answer me i couldn't be more apprecuated thanks
> > 
> > 
> > I can run the script on one gpu device, could you provide the runtime error info in detail?
> 
> i am in class. But i do experiment on windows ,did you do experiment on linux?

yes, mine is linux

**Comment by island-369** [comment]
I just changed the save format of the DIRE images to PNG, and it can no longer correctly classify real images.
I don't know the reason for this. Could it be that if the original image is JPG and the DIRE image is saved as PNG, some information is lost?

**Comment by lcayvinliu** [comment]
You can refer to these few references about dire problem：
1.
https://openaccess.thecvf.com/content/CVPR2024/html/Cazenavette_FakeInversion_Learning_to_Detect_Images_from_Unseen_Text-to-Image_Models_by_CVPR_2024_paper.html
2. http://arxiv.org/abs/2401.17879
in short, the DIRE results are all false (as noted in above papers). They
unfortunately preprocessed their data in such a way that all their “real”
DIRE images were JPEG compressed while all the “fake” DIRE images were
saved cleanly. So their model just learned to detect JPEG artifacts,
explaining the 100% accuracy on all their test sets…

**Comment by island-369** [comment]
> You can refer to these few references about dire problem：
> 1.
> https://openaccess.thecvf.com/content/CVPR2024/html/Cazenavette_FakeInversion_Learning_to_Detect_Images_from_Unseen_Text-to-Image_Models_by_CVPR_2024_paper.html
> 2. http://arxiv.org/abs/2401.17879
> in short, the DIRE results are all false (as noted in above papers). They
> unfortunately preprocessed their data in such a way that all their “real”
> DIRE images were JPEG compressed while all the “fake” DIRE images were
> saved cleanly. So their model just learned to detect JPEG artifacts,
> explaining the 100% accuracy on all their test sets…
> […](#)

If, in the author's code, the computed dire = torch.abs(imgs - recons) is used directly for classification instead of generating DIRE images, would it still be able to classify successfully?

and I noticed that in AI image detection datasets, real images are often stored in JPEG format, while AI-generated images are stored in PNG format. Could other detectors also be learning to distinguish JPEG and PNG images rather than identifying AI-generated and real images?

**Comment by lcayvinliu** [comment]
some detectors notice this jpeg/png problem, often use  jpeg quality=[70,
95 ] to compress the image in their training datasets， some papers show
this method can relieve the problem
in my practic, the data augment method,  mentioned in paper:
http://arxiv.org/abs/2406.19435, is effective.

**Comment by island-369** [comment]
> some detectors notice this jpeg/png problem, often use  jpeg quality=[70,
> 95 ] to compress the image in their training datasets， some papers show
> this method can relieve the problem
> in my practic, the data augment method,  mentioned in paper:
> http://arxiv.org/abs/2406.19435, is effective.
> […](#)

I noticed that in the paper 《A Sanity Check for AI-generated Image Detection》, they tested DIRE.  DIRE should classify all PNG images as AI-generated and all JPG images as real images. However, the test results in the paper seem to be different. Could it be that they made a mistake in their testing?
